sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,519 @@
+    /// <param name="loggedHeaderNames"></param>
+    /// <param name="loggedQueryParameters"></param>
+    /// <param name="requestIdHeaderName"></param>
+    /// <param name="logContent"></param>
+    /// <param name="maxLength"></param>
+    /// <param name="assemblyName"></param>
+    public ClientLoggingPolicy(
+        bool isLoggingEnabled = true,
+        List<string>? loggedHeaderNames = default,
+        List<string>? loggedQueryParameters = default,
+        string? requestIdHeaderName = default,
+        bool logContent = false,
+        int maxLength = 4 * 1024,
+        string? assemblyName = default)
+    {
+        LoggedHeaderNames = loggedHeaderNames ?? new List<string>();
+        LoggedQueryParameters = loggedQueryParameters ?? new List<string>();
+        _logContent = logContent;
+        _maxLength = maxLength;
+        _assemblyName = assemblyName;
+        _clientRequestIdHeaderName = requestIdHeaderName;
+        _isLoggingEnabled = isLoggingEnabled;
+    }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    public static ClientLoggingPolicy Default { get; } = new ClientLoggingPolicy();
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    public List<string> LoggedHeaderNames { get; internal set; }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    public List<string> LoggedQueryParameters { get; internal set; }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    internal PipelineMessageSanitizer? Sanitizer { get; set; }
+
+    /// <inheritdoc/>
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        if (!_isLoggingEnabled)
+        {
+            ProcessNext(message, pipeline, currentIndex);
+            return;
+        }
+
+        ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+    }
+
+    /// <inheritdoc/>
+    public override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        if (!_isLoggingEnabled)
+        {
+            await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+            return;
+        }
+
+        await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+    }
+
+    internal virtual void LogRequest()
+    {
+        if (!s_eventSource.IsEnabled())
+        {
+            return;
+        }
+    }
+
+    internal virtual void LogResponse()
+    {
+        if (!s_eventSource.IsEnabled())
+        {
+            return;
+        }
+    }
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        PipelineRequest request = message.Request;
+        string requestId = string.IsNullOrEmpty(_clientRequestIdHeaderName) ? string.Empty : GetRequestIdFromHeaders(request.Headers, _clientRequestIdHeaderName!);
```

Comment:
I just noticed, wrt `_clientRequestIdHeaderName!` ... we're using the null-forgiving operator `!` a lot in this code, but ideally, we'd only ever use this where we have 100% confidence that the value cannot ever be null.  If we know `_clientRequestIdHeaderName` can't be null here, would it make sense to declare the field as non-nullable?  Where does it get set?  If there's a good reason, it might be preferrable to make null part of the semantics rather than suppressing it.  Happy to sync on best practices or provide suggestions around this if that's of interest.

Reply:
For this one I just used the null-forgiving operator since I have a null check before using it, but I'll look at the alternatives you suggested so that this code better fits the best practices.